code/modules/holodeck/computer.dm

```diff
@@ -26,43 +26,50 @@
 	active_power_usage = 50
 
 	var/area/holodeck/linked
-	var/area/holodeck/program
-	var/area/holodeck/last_program
-	var/area/offline_program = /area/holodeck/rec_center/offline
+	var/area/holodeck/loaded
+	var/program = "holodeck_offline"//should these still be area vars after refactoring? what else can i use?
+	var/last_program
+	var/offline_program = "holodeck_offline"
 
 	var/list/program_cache
 	var/list/emag_programs
 
 	// Splitting this up allows two holodecks of the same size
 	// to use the same source patterns.  Y'know, if you want to.
-	var/holodeck_type = /area/holodeck/rec_center	// locate(this) to get the target holodeck
-	var/program_type = /area/holodeck/rec_center	// subtypes of this (but not this itself) are loadable programs
+	var/holodeck_type = /datum/map_template/holodeck
+	var/program_type = /datum/map_template/holodeck	// subtypes of this (but not this itself) are loadable programs
 
 	var/active = FALSE
 	var/damaged = FALSE
 	var/list/spawned = list()
 	var/list/effects = list()
 	var/current_cd = 0
+	var/datum/map_template/holodeck/template
+	var/turf/spawn_tile
+	var/turf/bottom_left
+	var/list/before_load = list()
+	var/list/after_load = list()
 
 /obj/machinery/computer/holodeck/Initialize(mapload)
 	..()
 	return INITIALIZE_HINT_LATELOAD
 
-/obj/machinery/computer/holodeck/LateInitialize()
-	if(ispath(holodeck_type, /area))
-		linked = pop(get_areas(holodeck_type, FALSE))
-	if(ispath(offline_program, /area))
-		offline_program = pop(get_areas(offline_program), FALSE)
-	// the following is necessary for power reasons
-	if(!linked || !offline_program)
-		log_world("No matching holodeck area found")
-		qdel(src)
-		return
+/obj/machinery/computer/holodeck/LateInitialize()//from here linked is populated and the program list is generated. its also set to load the offline program
+	linked = GLOB.areas_by_type[/area/holodeck/rec_center] //this should make current_area be the actual holodeck offline area object
+	bottom_left = locate(linked.x, linked.y, 2)
 	var/area/AS = get_area(src)
+
 	if(istype(AS, /area/holodeck))
 		log_mapping("Holodeck computer cannot be in a holodeck, This would cause circular power dependency.")
 		qdel(src)
 		return
+
+	// the following is necessary for power reasons
+	if(!linked || !offline_program)
+		log_world("No matching holodeck area found")
+		qdel(src)
+		return
+
```

Comment:
I think that `!offline_program` must be in separated check and log something like `Wrong offline_program variable`

Reply:
fixed